bitcoin/src/bip32.rs

```diff
@@ -257,6 +294,162 @@ impl serde::Serialize for ChildNumber {
+
+    /// Returns derivation path for a master key (i.e. empty derivation path)
+    pub fn master() -> NormalDerivationPath { NormalDerivationPath(vec![]) }
+
+    /// Returns whether derivation path represents master key (i.e. it's length
+    /// is empty). True for `m` path.
+    pub fn is_master(&self) -> bool { self.0.is_empty() }
+
+    /// Create a new [NormalDerivationPath] that is a child of this one.
+    pub fn child(&self, cn: NormalChildNumber) -> NormalDerivationPath {
+        let mut path = self.0.clone();
+        path.push(cn);
+        NormalDerivationPath(path)
+    }
+
+    /// Convert into a [NormalDerivationPath] that is a child of this one.
+    pub fn into_child(self, cn: NormalChildNumber) -> NormalDerivationPath {
+        let mut path = self.0;
+        path.push(cn);
+        NormalDerivationPath(path)
+    }
+    /// Concatenate `self` with `path` and return the resulting new path.
+    ///
+    /// ```
+    /// use bitcoin::bip32::{NormalDerivationPath, NormalChildNumber};
+    /// use std::str::FromStr;
+    ///
+    /// let base = NormalDerivationPath::from_str("m/42").unwrap();
+    ///
+    /// let deriv_1 = base.extend(NormalDerivationPath::from_str("0/1").unwrap());
+    /// let deriv_2 = base.extend(&[
+    ///     NormalChildNumber::ZERO,
+    ///     NormalChildNumber::ONE
+    /// ]);
+    ///
+    /// assert_eq!(deriv_1, deriv_2);
+    /// ```
+    pub fn extend<T: AsRef<[NormalChildNumber]>>(&self, path: T) -> NormalDerivationPath {
+        let mut new_path = self.clone();
+        new_path.0.extend_from_slice(path.as_ref());
+        new_path
+    }
+
+    /// Returns the `NormalDerivationPath` as a vector of u32 integers.
+    /// Unhardened elements are copied as is.
+    ///
+    /// ```
+    /// use bitcoin::bip32::NormalDerivationPath;
+    /// use std::str::FromStr;
+    ///
+    /// let path = NormalDerivationPath::from_str("m/84/0/0/0/1").unwrap();
+    /// const NORMAL: u32 = 0;
+    /// assert_eq!(path.to_u32_vec(), vec![84 + NORMAL, NORMAL, NORMAL, 0, 1]);
+    /// ```
+    pub fn to_u32_vec(self) -> Vec<u32> { self.into_iter().map(|num| num.0).collect() }
+
+    /// Creates a derivation path from a slice of u32s.
+    /// ```
+    /// use bitcoin::bip32::NormalDerivationPath;
+    ///
+    /// const NORMAL: u32 = 0;
+    /// let expected = vec![84 + NORMAL, NORMAL, NORMAL, 0, 1];
+    /// let path = NormalDerivationPath::from_u32_slice(expected.as_slice());
+    /// assert_eq!(path.to_u32_vec(), expected);
+    /// ```
+    pub fn from_u32_slice(numbers: &[u32]) -> Self {
+        numbers.iter().map(|&n| NormalChildNumber(n)).collect()
+    }
+}
+impl AsRef<[NormalChildNumber]> for NormalDerivationPath {
+    fn as_ref(&self) -> &[NormalChildNumber] { &self.0 }
+}
+
+impl From<NormalDerivationPath> for DerivationPath {
+    fn from(path: NormalDerivationPath) -> Self {
+        path.0.iter().map(|n| ChildNumber::from(*n)).collect()
+    }
+}
```

Comment:
Maybe, casting directly instead of from ?

Reply:
We can't do that because they have different layouts. I've suggested silently changing the layout but that has other implications, so I don't really know which is better.

Reply:
I can do just a n.into() here

Reply:
Oh, you can, it doesn't matter those are the same . I though you mean casting using `as` (or casting the pointers).